wsnet/dial.go

```diff
@@ -173,6 +173,19 @@ func (d *Dialer) negotiate() (err error) {
 		})
 	}()
 
+	go func() {
+		// If a connection is opened but the other end may not be, negotiation
+		// can get stuck forever. We don't want this, so we must listen to the
+		// context as well.
+		<-ctx.Done()
+		select {
+		case <-errCh:
+		default:
+			errCh <- ctx.Err()
+			close(errCh)
```

Comment:
You're closing this channel in multiple goroutines, it's going to panic on the second close attempt. You should consider not closing it, or making some `sendErr` function that wraps these two lines in a sync.Once

Reply:
This switch checks if the channel is open... so it shouldn't close twice.

Reply:
Err I guess it would in the other one. I'll just check if it's open there too.